frontend/src/modules/editor/themeEditor/views/editorThemingView.js

```diff
@@ -0,0 +1,560 @@
+        var preset = this.getSelectedPreset();
+        var baseSettings = (preset) ? preset.get('properties') : this.getDefaultThemeSettings();
+        shouldShow = false;
+        for (var key in currentSettings) {
+          if (currentSettings[key] && baseSettings[key]) {
+            // Check for nested properties
+            if (typeof currentSettings[key] === 'object') {
+              for (var innerKey in currentSettings[key]) {
+                if (currentSettings[key][innerKey].toString() !== baseSettings[key][innerKey].toString()) {
+                  shouldShow = true;
+                }
+              }
+            } else if (currentSettings[key].toString() !== baseSettings[key].toString()) {
+              shouldShow = true;
+            }
+          }
+        }
+      }
+      var $reset = $('.editor-theming-sidebar-reset');
+      shouldShow ? $reset.css('visibility', 'visible') : $reset.css('visibility', 'hidden');
+    },
+
+    getPresetSelection: function() {
+      return this.settings.presetSelection;
+    },
+
+    setPresetSelection: function(id) {
+      this.settings.presetSelection = id;
+    },
+
+    /**
+    * Event handling
+    */
+
+    onEditPreset: function(data) {
+      var model = this.presets.findWhere({ displayName: data.oldValue });
+      model.set('displayName', data.newValue);
+      model.save();
+    },
+
+    onDeletePreset: function(preset) {
+      var toDestroy = this.presets.findWhere({ displayName: preset });
+      this.removePresetOption(toDestroy.get('_id'));
+      toDestroy.destroy();
+    },
+
+    removePresetOption: function(id) {
+      var select = this.$('.preset select');
+      if (select.val() === id) {
+          select.val('');
+      }
+      select.find('option[value="' + id + '"]').remove();
+    },
+
+    onCollectionReady: function(collection) {
+      if (collection === this.themes || collection === this.presets) {
+        collection.ready = true;
+        if (this.isDataLoaded()) this.trigger('dataReady');
+      }
+      // must just be a model
+      else {
+        this.updateSelects();
+      }
+    },
+
+    onError: function(collection, response, options) {
+      Origin.Notify.alert({
+        type: 'error',
+        text: response
+      });
+    },
+
+    onThemeChanged: function() {
+      this.setPresetSelection(null);
+      this.updatePresetSelect();
+      this.renderForm();
+      this.updateRestorePresetButton(false);
+    },
+
+    onPresetChanged: function(event) {
+      var preset = this.presets.findWhere({ _id: $(event.currentTarget).val() });
+      var settings = preset && preset.get('properties') || this.getDefaultThemeSettings();
+      this.setPresetSelection($(event.currentTarget).val());
+      this.restoreFormSettings(settings);
+      this.updateRestorePresetButton(false);
+    },
+
+    onFieldChanged: function() {
+        this.updateRestorePresetButton(true);
```

Comment:
I'm seeing that when you change a colour and then put it back to the preset's value, the restore to preset button is still showing when it shouldn't.

It should be as simple as removing the parameter from this function call but the code doesn't seem to be properly calculating the current selections?

Btw, completely agree with you re: the save button.

Reply:
Yes you're right, the code seems to be comparing what is currently saved against the preset. It should be comparing what is currently selected in the form against the preset.

Looking in to how to get what is currently selected in the form. If you know how to to this, or you know of somewhere else in the authoring tool that might be doing something similar, then please let me know because it doesn't seem straightforward to me at the moment.

Reply:
Reverse-engineering from the way values are set in the current code, this sort of thing will work for unnested properties:
```js
getCurrentSettings: function() {
  if (!this.form) {
    return Origin.editor.data.course.get('themeVariables');
  }

  return _.mapObject(this.form.fields, function(field) {
    return field.getValue();
  });
},
```

The get functions (above and `getDefaultThemeSettings()`) should probably be modified to flatten any nested properties so that it's simpler to compare.